client/src/_transport_layer.rs

```diff
@@ -0,0 +1,257 @@
+            .for_each(move |forwarded| {
+                let myself = self.clone();
+                async move {
+                    myself.dispatch(forwarded).await;
+                }
+            })
+            .await
+    }
+
+    async fn forward_channel(
+        &self,
+        node_id: NodeId,
+        channel: TransportType,
+    ) -> Option<ForwardSender> {
+        let state = self.state.read().await;
+        match channel {
+            TransportType::Reliable => state.forward_reliable.get(&node_id).cloned(),
+            TransportType::Transfer => state.forward_transfer.get(&node_id).cloned(),
+            TransportType::Unreliable => state.forward_unreliable.get(&node_id).cloned(),
+        }
+    }
+
+    async fn set_forward_channel(
+        &self,
+        node_id: NodeId,
+        channel: TransportType,
+        tx: ForwardSender,
+    ) {
+        let mut state = self.state.write().await;
+        match channel {
+            TransportType::Reliable => state.forward_reliable.insert(node_id, tx),
+            TransportType::Transfer => state.forward_transfer.insert(node_id, tx),
+            TransportType::Unreliable => state.forward_unreliable.insert(node_id, tx),
+        };
+    }
+
+    pub async fn forward(&self, node_id: NodeId) -> anyhow::Result<ForwardSender> {
+        self.forward_generic(node_id, TransportType::Reliable).await
+    }
+
+    pub async fn forward_transfer(&self, node_id: NodeId) -> anyhow::Result<ForwardSender> {
+        self.forward_generic(node_id, TransportType::Transfer).await
+    }
+
+    /// NodeId can be either default or secondary.
+    /// TODO: Make this function resistant to dropping future
+    pub async fn forward_generic(
+        &self,
+        node_id: NodeId,
+        channel: TransportType,
+    ) -> anyhow::Result<ForwardSender> {
+        match self.forward_channel(node_id, channel).await {
+            // If connection was closed in the meantime, it will be initialized on demand.
+            // It will be problematic in some cases, because this can last up to a few seconds.
+            // In worst case scenario initialization will fail and we will wait 5s until timeout.
+            // Since user uses channel, sending will return immediately after item will be taken from
+            // queue, so he won't find out, but the response he expects won't come.
+            // This is argument for changing channels API to `TcpSender`.
+            Some(tx) => Ok(tx),
+            None => {
+                // Check if this isn't secondary identity. TcpLayer should always get default id.
+                // TODO: Consider how to handle changing identities.
+                let info = self.session_layer.query_node_info(node_id).await?;
+
+                if let Some(tx) = self.forward_channel(info.node_id(), channel).await {
+                    self.set_forward_channel(node_id, channel, tx.clone()).await;
+                    return Ok(tx);
+                }
```

Comment:
```suggestion
                if let Some(tx) = self.get_forward_channel_for_secondary_identity(node_id, channel) {
                    self.set_forward_channel(node_id, channel, tx.clone()).await;
                    return Ok(tx);
                }
```

Reply:
OR it could be moved totally inside tcp_layer so here code will call only `session_layer.session(...)` as in unreliable case